gateway-service/src/main/java/org/zowe/apiml/gateway/security/config/NewSecurityConfiguration.java

```diff
@@ -0,0 +1,394 @@
+import org.springframework.security.web.authentication.logout.LogoutHandler;
+import org.springframework.security.web.firewall.StrictHttpFirewall;
+import org.springframework.security.web.util.matcher.RegexRequestMatcher;
+import org.zowe.apiml.gateway.controllers.AuthController;
+import org.zowe.apiml.gateway.controllers.CacheServiceController;
+import org.zowe.apiml.gateway.error.InternalServerErrorController;
+import org.zowe.apiml.gateway.security.login.x509.X509AuthenticationProvider;
+import org.zowe.apiml.gateway.security.query.*;
+import org.zowe.apiml.gateway.security.service.AuthenticationService;
+import org.zowe.apiml.gateway.security.ticket.SuccessfulTicketHandler;
+import org.zowe.apiml.gateway.services.ServicesInfoController;
+import org.zowe.apiml.security.common.config.AuthConfigurationProperties;
+import org.zowe.apiml.security.common.config.HandlerInitializer;
+import org.zowe.apiml.security.common.content.BasicContentFilter;
+import org.zowe.apiml.security.common.content.CookieContentFilter;
+import org.zowe.apiml.security.common.handler.FailedAuthenticationHandler;
+import org.zowe.apiml.security.common.login.LoginFilter;
+import org.zowe.apiml.security.common.login.ShouldBeAlreadyAuthenticatedFilter;
+
+import java.util.Set;
+
+/**
+ * Main configuration place for Gateway endpoint security
+ *
+ * Security is configured with separate filterchains per groups of endpoints
+ * The main theme is to keep the filterchains independent and isolated
+ * Gives more control over the behavior of individual filter chain
+ * This makes the security filters simpler as they don't have to be smart
+ * Also separates the x509 filters per filterchain for more control
+ *
+ * Authentication providers are initialized per filterchain's needs. No unused auth providers on chains.
+ */
+
+@ConditionalOnProperty(name = "apiml.security.filterChainConfiguration", havingValue = "new", matchIfMissing = false)
+@Configuration
+@EnableWebSecurity
+@RequiredArgsConstructor
+@Slf4j
+public class NewSecurityConfiguration {
+
+    private String applicationContextPath = "/gateway"; //NOSONAR this is hardcoded as there is no value for this in config
+
+    private static final String EXTRACT_USER_PRINCIPAL_FROM_COMMON_NAME = "CN=(.*?)(?:,|$)";
+
+    private final ObjectMapper securityObjectMapper;
+    private final AuthenticationService authenticationService;
+    private final AuthConfigurationProperties authConfigurationProperties;
+    private final HandlerInitializer handlerInitializer;
+    private final SuccessfulQueryHandler successfulQueryHandler;
+    private final SuccessfulTicketHandler successfulTicketHandler;
+    @Qualifier("publicKeyCertificatesBase64")
+    private final Set<String> publicKeyCertificatesBase64;
+    private final X509AuthenticationProvider x509AuthenticationProvider;
+
+
+    @Configuration
+    @RequiredArgsConstructor
+    @Order(1)
+    class AuthenticationFunctionality extends WebSecurityConfigurerAdapter {

```

Comment:
Could you please add comment to the whole class on what this one should do?

Reply:
How detailed description do you look for? The class is named to suggest it serves the auth functionality. Do you need more details or do you want to describe how the filter is constructed? There are some comments inline to make reading easier..

Reply:
I would like to have a Javadoc in the same way as for the others outlining which scenarios are handled

Reply:
Ok, pushed it now